gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java

```diff
@@ -29,8 +29,7 @@ public HealthCheckObserver(RoutingManager routingManager)
     public void observe(java.util.List<ClusterStats> clustersStats)
     {
         for (ClusterStats clusterStats : clustersStats) {
-            routingManager.upateBackEndHealth(clusterStats.getClusterId(), clusterStats.isHealthy());
-            routingManager.updateBackEndHealthDB(clusterStats);
```

Comment:
Why did you remove this line?

Reply:
this is the line that caused the problem. It directly manipulates the DB, which is unwanted. This method has been deleted in `routingManager`

Reply:
> having multiple gateways could lead to inconsistent states for sure

In deploying using Helm Chart and for HA purpose for gateway server itself, people would use multiple gateways. Our team too are using HPA for the gateway itself, with 3 in minimum. 

If the status is local with serveral servers, then everytime I refresh the page I could get different Healthy status right?

I think a consistent sync is necessary, like 
- if status is inactive then the monitor could query and check if it is healthy
  - if the result is unhealthy, show unhealthy (meaning that it will do health check later) 
  - if the result is pending, show pending (similar to unhealthy but pending could also mean it is in checking state?)
  - if the result is healthy, then update DB so no other gateway server linked to same db would do unnecessary querying
- if status is active, then monitor does nothing

Reply:
by status do you mean backend status? I think you have active and inactive cases reversed..?

Reply:
sry for confusion. status meant by active field in db

Reply:
Technically if you refresh the page you shouldn't get different results because if the pod is up and running, the healthcheck should be performed automatically. Therefore, you should get fairly consistent result given that the clusters are not flaky.

However, there will still be inconsistent states if the clusters are flaky and the gateway is not picking up yet. 

I see the value of having a central place to store this info to avoid inconsistent states. While this is great, I think it might lead to some kind of race condition, and could potentially cause requests to be routed to an unhealthy cluster. 

For example, if gateway A sees cluster A is bad and tries to mark it unhealthy, in the meantime gateway B looks at DB and thinks cluster A is healthy, and hence routes requests to cluster A. 

If this information is only available locally, then each gateway will manage its own clusters' states. Then there won't be any race condition.

Any thoughts?

Reply:
> Technically if you refresh the page you shouldn't get different results because if the pod is up and running, the healthcheck should be performed automatically.  Therefore, you should get fairly consistent results given that the clusters are not flaky.

True, this does not happen that often. 

> For example, if gateway A sees cluster A is bad and tries to mark it unhealthy, in the meantime gateway B looks at DB and thinks cluster A is healthy, and hence routes requests to cluster A.

If the DB has value, i don't see any cases where gateway A, B sees cluster A active status differently.

Or maybe have a lock field? 
Assuming cluster A is inactive, and cluster B active.
- Gateway A scans if there is any inactive cluster. Finds cluster A, check for lock, if there isn't any lock, locks cluster A
  - Update (querying to see number of nodes) and releases lock afteward.
- Gateway B, also scans, and checks for lock. Cluster A is locked so skip.

Reply:
the current behavior is that if any cluster is **inactive** (meaning it's switched off manually), then the healthcheck will not be performed, since it doesn't make sense to perform healthcheck on **inactive** cluster.

If the cluster is **active**, then the healthcheck will perform. The cluster could be healthy, pending, or unhealthy. 

In your case, if Gateway A finds any inactive clusters, should skip them. It should only check on active clusters.

Reply:
The issue with storing healthcheck states in DB is that the healthcheck is done every minute and it's supposed to be a fast/small ping test. If we store this result in DB, we will need to look at DB everytime when healthcheck is performed, and hence it's adding more overhead. Also, when routingManager is choosing which cluster to route, it will need to look at DB to ensure the cluster is actually healthy. This could slow down the operation time.

Reply:
I'll start a Slack thread to discuss design options